// assistant-stream@0.2.23/dist/core/serialization/data-stream/chunk-types.js
// src/core/serialization/data-stream/chunk-types.ts
var DataStreamStreamChunkType = /* @__PURE__ */ ((DataStreamStreamChunkType2) => {
  DataStreamStreamChunkType2["TextDelta"] = "0";
  DataStreamStreamChunkType2["Data"] = "2";
  DataStreamStreamChunkType2["Error"] = "3";
  DataStreamStreamChunkType2["Annotation"] = "8";
  DataStreamStreamChunkType2["ToolCall"] = "9";
  DataStreamStreamChunkType2["ToolCallResult"] = "a";
  DataStreamStreamChunkType2["StartToolCall"] = "b";
  DataStreamStreamChunkType2["ToolCallArgsTextDelta"] = "c";
  DataStreamStreamChunkType2["FinishMessage"] = "d";
  DataStreamStreamChunkType2["FinishStep"] = "e";
  DataStreamStreamChunkType2["StartStep"] = "f";
  DataStreamStreamChunkType2["ReasoningDelta"] = "g";
  DataStreamStreamChunkType2["Source"] = "h";
  DataStreamStreamChunkType2["RedactedReasoning"] = "i";
  DataStreamStreamChunkType2["ReasoningSignature"] = "j";
  DataStreamStreamChunkType2["File"] = "k";
  DataStreamStreamChunkType2["AuiUpdateStateOperations"] = "aui-state";
  DataStreamStreamChunkType2["AuiTextDelta"] = "aui-text-delta";
  DataStreamStreamChunkType2["AuiReasoningDelta"] = "aui-reasoning-delta";
  return DataStreamStreamChunkType2;
})(DataStreamStreamChunkType || {});
export {
  DataStreamStreamChunkType
};
